materialize-firebolt/transactor.go

```diff
@@ -0,0 +1,287 @@
+		var propName = spec.FieldSelection.Values[i]
+
+		if raw, ok := value.([]byte); ok {
+			document[propName] = json.RawMessage(raw)
+		} else {
+			document[propName] = value
+		}
+	}
+
+	jsonDoc, err := json.Marshal(document)
+	if err != nil {
+		return nil, fmt.Errorf("failed to serialize the addition document: %w", err)
+	}
+
+	return jsonDoc, nil
+}
+
+// write file to S3 and keep a record of them to be copied to the table
+func (t *transactor) Store(it *pm.StoreIterator) error {
+	awsConfig := aws.Config{
+		Credentials: credentials.NewStaticCredentials(t.awsKeyId, t.awsSecretKey, ""),
+		Region:      &t.awsRegion,
+	}
+	sess := session.Must(session.NewSession(&awsConfig))
+	uploader := s3manager.NewUploader(sess)
+
+	uploads := []s3manager.BatchUploadObject{}
+
+	files := make(map[int][]byte)
+
+	for it.Next() {
+		doc, err := t.projectDocument(t.bindings[it.Binding].spec, it.Key, it.Values)
+		if err != nil {
+			return fmt.Errorf("projecting new store json: %w", err)
+		}
+
+		if _, ok := files[it.Binding]; ok {
+			files[it.Binding] = append(files[it.Binding], byte('\n'))
+			files[it.Binding] = append(files[it.Binding], doc...)
+		} else {
+			files[it.Binding] = doc
+		}
+	}
+
+	for bindingIndex, data := range files {
+		cp := t.checkpoint.Files[bindingIndex]
+
+		uploads = append(uploads, s3manager.BatchUploadObject{
+			Object: &s3manager.UploadInput{
+				Bucket: aws.String(cp.Bucket),
+				Key:    aws.String(cp.Key),
+				Body:   bytes.NewReader(data),
+			},
+		})
+	}
+
+	uploadIterator := s3manager.UploadObjectsIterator{
+		Objects: uploads,
+	}
+
+	err := uploader.UploadWithIterator(aws.BackgroundContext(), &uploadIterator)
+	if err != nil {
+		return fmt.Errorf("uploading files to s3: %w", err)
+	}
+
+	return nil
+}
+
+func (t *transactor) Commit(ctx context.Context) error {
+	return nil
+}
+
+// load stored file into table
+func (t *transactor) Acknowledge(context.Context) error {
+
+	if len(t.checkpoint.Files) < 1 {
+		return nil
+	}
+
+	sourceFileNames := map[int][]string{}
+	for _, item := range t.checkpoint.Files {
+		binding := item.Binding
+		fileName := item.Key
+
+		if _, ok := sourceFileNames[binding]; ok {
+			sourceFileNames[binding] = append(sourceFileNames[binding], fileName)
+		} else {
+			sourceFileNames[binding] = []string{fileName}
+		}
+	}
+
+	// Acknowledge step might end up being run by a separate process
+	// different from the original, and the new transactor process might be initialised with
+	// a new version of the MaterializationSpec. This means it's possible to have an Acknowledge
+	// with MaterializationSpec v1 be run by a transactor initialised with MaterializationSpec v2.
+	// As such, we keep a copy of the queries generated from the MaterializationSpec of the
+	// transaction to which this Acknowledge belongs in the checkpoint to avoid inconsistencies.
+	for binding, fileNames := range sourceFileNames {
+		insertQuery := t.checkpoint.Queries.Bindings[binding].InsertFromTable
```

Comment:
Since we're stuffing queries into each checkpoint anyway, what do you think about pre-filling in these values so that the checkpoint stores completed query literals? Then Acknowledge can side-step needing to know anything about bindings.

Also, sanity checking, what happens if a specific load query is run twice? The second load is a no-op?

Reply:
Running a query twice loads the data twice (creates duplicate rows), is that a case we need to handle to make sure it's no-op?

Reply:
Yes, definitely It's expected that we'll regularly see Acknowledge run twice for a given commit: a prior connector completes a transaction, acknowledges it, and then exits. A subsequent connector then re-acknowledges that same transaction as its first action.

If the second acknowledge loads data twice, that's a violation of exactly-once semantics.

Worse, it's possible that an Acknowledge can _partially_ complete (only _some_ of the queries are issued), then perhaps crash while awaiting one of these running INSERT queries. A subsequent invocation then tries the whole Acknowledge again.

The absolute ideal would be to have each individual query be idempotent and atomic -- we can send a single query to firebolt, then crash without getting a response back, and it will either a) complete such that sending the query again is a no-op, or b) fail with no side-effects.

Not sure if or what is achievable w/in Firebolt though. Is there a `CASE WHEN` equivalent ?

Reply:
I'll look into this to see how can we make this an idempotent operation 👍🏽

Reply:
This should make the insertion idempotent: https://github.com/estuary/flow/pull/407/commits/cd715d1be1b9f99608d3ec8d8f93bd29f5cf9efd

Reply:
I just tried it by running this code (basically running each query twice)
```
	for _, query := range t.checkpoint.Queries {
		_, err := t.fb.Query(query)
		if err != nil {
			return fmt.Errorf("moving files from external to main table: %w", err)
		}
		_, err = t.fb.Query(query)
		if err != nil {
			return fmt.Errorf("moving files from external to main table: %w", err)
		}
	}
```

No duplicates are inserted 👍🏽